### ISSUE TITLE
Math brute force failures on aarch64

### ISSUE THREAD

**Issue by julienhascoet** [KhronosGroup/OpenCL-CTS#1233]
Dear all,

On x86_64 (Ubuntu distro 1804) math brute force is working properly on my device, however on aarch64 (Ubuntu distro 1804) I get errors on the following subtests:

> acosh...
>   3:           acosh fp32     ................passed        1.92 @ 0x1.07a78p+0
>   4:           acosh fp64     ...........
> ERROR: acosh: -7.000000 ulp error at 0x1.0100000000000p+2 (0x4010100000000000): *0x1.08a2a10b57698p+1 vs. 0x1.69ebc1b1da599p-4
> .acosh FAILED

> cbrt...
>  23:            cbrt fp32     ................passed        3.52 @ -0x1.ec628p-7
>  24:            cbrt fp64     
> ERROR: cbrt: 59.119652 ulp error at -0x1.0000000000000p+1019 (0xffa0000000000000): *-0x1.965fea53d6ddep+339 vs. -0x1.0000000000000p+341
> ERROR: cbrt: -95.345238 ulp error at -0x1.0000000000000p+1019 (0xffa0000000000000): *-0x1.965fea53d6ddep+339 vs. -0x1.965fea53d6e3dp+337
> ERROR: cbrt: 58.772907 ulp error at -0x1.0000000000000p+1019 (0xffa0000000000000): *-0x1.965fea53d6ddep+339 vs. -0x1.0000000000000p+339
> ERROR: cbrt: 58.426163 ulp error at -0x1.0000000000000p+1019 (0xffa0000000000000): *-0x1.965fea53d6ddep+339 vs. -0x1.0000000000000p+337
> ERROR: cbrt: 73.858772 ulp error at -0x1.00000000000c0p+1019 (0xffa00000000000c0): *-0x1.965fea53d6ea2p+339 vs. -0x1.428a2f98d72dcp+340
> ERROR: cbrt: -94.794823 ulp error at -0x1.0000000000000p+1019 (0xffa0000000000000): *-0x1.965fea53d6ddep+339 vs. -0x1.965fea53d6e3dp+339
> ERROR: cbrt: 73.421906 ulp error at -0x1.00000000000c0p+1019 (0xffa00000000000c0): *-0x1.965fea53d6ea2p+339 vs. -0x1.428a2f98d72dcp+338
> ERROR: cbrt: 72.985039 ulp error at -0x1.00000000000c0p+1019 (0xffa00000000000c0): *-0x1.965fea53d6ea2p+339 vs. -0x1.428a2f98d72dcp+336
> cbrt FAILED

> log...
> Skipping reduced precision testing for device with version 1.2 or less
> 97:             log fp32     ................passed        0.83 @ 0x1.471c8p-1
>  98:             log fp64     ...........
> ERROR: log: -3.210737 ulp error at 0x1.3fffffffff800p-5 (0x3fa3fffffffff800): *-0x1.9f0d4423f21fcp+1 vs. -0x1.e148a1a274067p-2
> .log FAILED

> log2...
> Skipping reduced precision testing for device with version 1.2 or less
>  99:            log2 fp32     ................passed        1.60 @ 0x1.68218p+0
> 100:            log2 fp64     ...........
> ERROR: log2: -3.700597 ulp error at 0x1.03ffffffffdc0p+1 (0x40003ffffffffdc0): *0x1.05b9e5a170818p+0 vs. -0x1.f48c34bd1efd4p-1
> .log2 FAILED

> log10...
> Skipping reduced precision testing for device with version 1.2 or less
> 101:           log10 fp32     ................passed        1.98 @ 0x1.4c2fp+0
> 102:           log10 fp64     ...........
> ERROR: log10: 3.290207 ulp error at 0x1.2400000000500p-3 (0x3fc2400000000500): *-0x1.b11ffa78a3865p-1 vs. -0x1.f37b15baaf995p-3
> .log10 FAILED

> pow...
> Skipping reduced precision testing for device with version 1.2 or less
> 119:             pow fp32     ................passed        0.50 @ {0x1.cc1844p-18, 0x1.3038ccp-4}
> 120:             pow fp64     
> ERROR: pow: -44.043446 ulp error at {0x1.063ad47c926fep+1006, 0x1.eb45d19738616p-3}: *0x1.411875d076d2bp+241 vs. 0x1.411875d076cffp+241
> ERROR: pow: -16.497797 ulp error at {0x1.1a8180c35b0fdp-848, -0x1.171889aa2f4c2p-4}: *0x1.b51e1dc4f67bep+57 vs. 0x1.b51e1dc4f67aep+57
> ERROR: pow: 29.791166 ulp error at {0x1.ede9f6ea89a4bp-6, 0x1.f2706f8b4df2bp+5}: *0x1.2eef28513ad9dp-315 vs. 0x1.2eef28513adbbp-315
> ERROR: pow: -121.332649 ulp error at {0x1.d49208126408ap+889, 0x1.7a6e029d4d2d0p-2}: *0x1.d1191c538634dp+328 vs. 0x1.d1191c53862d4p+328
> ERROR: pow: -21.694351 ulp error at {0x1.2604915ca5e77p-197, -0x1.61ff8eee3edd5p-2}: *0x1.0620b5e3b9428p+68 vs. 0x1.0620b5e3b9412p+68
> ERROR: pow: 37.226974 ulp error at {0x1.782245e7b2cb3p-591, -0x1.ec0613186b390p-4}: *0x1.e66689d54b1b6p+70 vs. 0x1.e66689d54b1dbp+70
> ERROR: pow: -35.822289 ulp error at {0x1.88056e2d3cc14p+719, -0x1.ecacc1b9cd94fp-3}: *0x1.d951fbe206a3dp-174 vs. 0x1.d951fbe206a19p-174
> ERROR: pow: -24.152452 ulp error at {0x1.baa1d6a827c6cp+797, 0x1.11770ede57733p-3}: *0x1.70ed55565a00ap+106 vs. 0x1.70ed555659ff2p+106
> pow FAILED

> powr...
> 123:            powr fp32     ................passed        0.50 @ {0x1.1d5512p+48, 0x1.4ec8bep+1}
> 124:            powr fp64     
> ERROR: powr: -108.708038 ulp error at {0x1.7b6e808beb450p-403, -0x1.0fdb876afaf15p+0}: *0x1.48a40a8926127p+427 vs. 0x1.48a40a89260bap+427
> ERROR: powr: -39.161736 ulp error at {0x1.42d72119394b4p-35, 0x1.f66f03b9003dap+0}: *0x1.f3a4033bf6503p-69 vs. 0x1.f3a4033bf64dcp-69
> ERROR: powr: 19.049768 ulp error at {0x1.4b4d4212884aep-521, 0x1.5308c687b02a5p-4}: *0x1.dfe3aed963b3dp-44 vs. 0x1.dfe3aed963b50p-44
> ERROR: powr: 43.080959 ulp error at {0x1.8744d60a6d023p+722, 0x1.c8bd42703bf3bp-2}: *0x1.3d55f3a32e853p+322 vs. 0x1.3d55f3a32e87ep+322
> ERROR: powr: 43.743641 ulp error at {0x1.ef9c16363c432p+107, -0x1.bac6968a297aep+0}: *0x1.37f34ea3bd076p-187 vs. 0x1.37f34ea3bd0a2p-187
> ERROR: powr: -157.203827 ulp error at {0x1.eca54fd9ccc63p+665, -0x1.4118bf70cf730p+0}: *0x1.a4b4ecab49f15p-836 vs. 0x1.a4b4ecab49e78p-836
> ERROR: powr: 18.505419 ulp error at {0x1.31cf03eab9d47p+699, 0x1.bab4a5e4fef26p-4}: *0x1.7dfa415c083b2p+75 vs. 0x1.7dfa415c083c5p+75
> ERROR: powr: 270.473236 ulp error at {0x1.c68ed5a58491fp-346, 0x1.c015402693329p+0}: *0x1.c980a6bfc3ad1p-605 vs. 0x1.c980a6bfc3bdfp-605
> powr FAILED

All other subtests are successful.

Are you aware of such differences between x86_64 and aarch64 ?

Thanks in advanced.

**Comment by julienhascoet** [comment]
Hi there,

could this issue be explained by https://github.com/KhronosGroup/OpenCL-CTS/pull/1255

@mantognini @EwanC any idea ?

Thanks in advance.

**Comment by EwanC** [comment]
> 
> 
> Hi there,
> 
> could this issue be explained by #1255
> 
> @mantognini @EwanC any idea ?
> 
> Thanks in advance.

One thing that jumps out is that it's only the double precision fp64 tests which are failing, rather than single precision fp32, so  #1255 wouldn't be my first thought. 

For `cbrt` looks like the same input `-0x1.0000000000000p+1019` is failing with different output values, this nondeterministic behaviour is also suspect.

**Comment by mantognini** [comment]
Indeed, this is suspicious. @julienhascoet, if you believe this is an issue with the CTS and not the device itself, it would be interesting to know which revision(s) you used to run the CTS, and whether other, older revisions are also affected.

**Comment by julienhascoet** [comment]
Thanks a lot, @mantognini I used:

```
commit b8d1ea9962a7bb443e845f5f035b254172fc4ae7
Author: Marco Antognini <marco.antognini@arm.com>
Date:   Wed Apr 21 10:45:34 2021 +0100
```
I did not have time to debug this already, however, having differences on ARM host only (it's working on x86), it makes me believe it's a CTS issue regarding references. But as it's complex, I'll need to find some time to understand what's happening on the ARM host.
